parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala

```diff
@@ -214,21 +214,27 @@ class OpTreeContext(parser: Expr[Parser])(using Quotes) {
     }
   }
 
-  def deconstruct[I <: HList: Type, O <: HList: Type](rule: Expr[Rule[I, O]]): OpTree = rule match {
-    case '{
```

Comment:
I think this pattern will not work in general. The problem is that the types are just not the right ones. The existing code assumes that lhs, rhs and also the result of the chaining all have the same type which is not true in general.

It does not yet fail any tests because all of the implemented rules so far are Rule0 but it will be a problem going forward.

The only way to match those combinator rules correctly is to go down to the term-level and match these expression that way (basically disregarding any type information while matching).

(Replicating [this comment](https://github.com/sirthias/parboiled2/commit/c545e66fb0be47f65bb8c90b4e714c9098671882#r55637143) on an outdated commit)